### ISSUE TITLE
Incoming mails are not detected as replies

### ISSUE THREAD

**Issue by The-Compiler** [django-helpdesk/django-helpdesk#756]
Hi! I'm only a user of django-helpdesk, but I've seen this with various instances of it (DjangoGirls and Europython 2019) so far.

I open a ticket and get a confirmation mail with a ticket ID. Since I have additional information to share, I reply to the mail, and I don't touch the subject. Here are the headers of my reply:

```
Date: Mon, 20 May 2019 13:13:08 +0200
From: Florian Bruhin <me@the-compiler.org>
To: EuroPython 2019 Helpdesk <helpdesk@europython.eu>
Subject: Re: Training ticket for accepted workshop / early-bird ticket
 (Opened) [ep2019-00001]
Message-ID: <20190520111308.jwvu5tzqgpbplxyn@hooch.localdomain>
References: <20190520074803.32736.24972@5db3fa438c48>
MIME-Version: 1.0
Content-Type: multipart/signed; micalg=pgp-sha256;
        protocol="application/pgp-signature"; boundary="y5kita2yim63hf4r"
Content-Disposition: inline
In-Reply-To: <20190520074803.32736.24972@5db3fa438c48>
User-Agent: NeoMutt/20180716
```

I then get an answer with a new ticket ID:

```
From: EuroPython 2019 Helpdesk <helpdesk@europython.eu>
To: me@the-compiler.org
Subject: Training ticket for accepted workshop / early-bird ticket (Opened) [ep2019-00001] (Opened) [ep2019-00002]

This is a courtesy e-mail to let you know that we have received your helpdesk query with a subject of "Training ticket for accepted workshop / early-bird ticket
 (Opened) [ep2019-00001]".

You do not have to do anything further at this stage. Your ticket has been assigned a number of [ep2019-00002] and will be responded to shortly.

If you wish to send us further details, or if you have any queries about this ticket, please include the ticket id of '[ep2019-00002]' in the subject. The easiest way to do this is just press "reply" to this message.
```

(ID's changed in this issue to avoid other people viewing my tickets - the only change is replacing the 5-digit numbers by 00001 and 00002)

**Comment by gwasser** [comment]
Ah yes that's definitely not working as designed. Is this on the latest release (0.2.x) or `develop`? Thanks for reporting!

**Comment by The-Compiler** [comment]
Sorry, "a user" was ambiguous here - I meant "someone sending emails to django-helpdesk instances", not "someone running django-helpdesk".

In this instance, apparently they use [their own fork](https://github.com/EuroPython/ep-helpdesk), so it's entirely possible the bug is actually there. However, I've seen this happen a couple of years ago in a different django-helpdesk instance (I wish I remembered where it was...).

Is there a demo instance with emails enabled somewhere I could use to test? If not, how hard would it be to set up my own with email support? How are incoming mails fetched?

Happy to try things out, but if it's more work than just piping an email or setting up IMAP/SMTP, I'm afraid I'll need to postpone this a bit, as I'm currently busy with my bachelor thesis for another month or so :wink:

**Comment by gwasser** [comment]
We unfortunately don't have a public demo instance with emails enabled. You could run the demo locally maybe in a VM with a basic email server running to test it out. Alternately I think you can use IMAP to log in and check Gmail, so you could use a free webmail account as your source to poll from a local testing instance.

No worries on time frame, I completely understand! I've been in a similar position for a few months and haven't been able to give the 0.3 release the bugfix spree needed for release just yet. Hopefully sometime soon! In the meantime definitely welcome pull requests and help, thanks!

**Comment by mattbuff-nynhp** [comment]
I can confirm that this bug occurs on the latest release (0.2.16) installed via pip.

**Comment by gwasser** [comment]
Thanks for the confirmation @mattbuff-nynhp! @The-Compiler have you had a chance to look at a fix? No worries, just wanted to follow up since the thread got a bump.

**Comment by The-Compiler** [comment]
Nope, sorry! Bachelor thesis is finished by now, but I have another exam coming up and generally a lot on my plate with [my own project](https://www.qutebrowser.org/) - I might be able to take a look in September/October or so, but can't promise anything.

**Comment by csinsota** [comment]
Could it be that the ticket was moved to a different queue inbetween?

The regex used for matching incoming emails to tickets includes the slug of the queue, but in subjects the current queue slug of the ticket is used.
There can be problems if you move a ticket to a different queue.

Steps to recreate (tested in master):
Create a queue with slug 'queue1', setup to fetch email, allow e-mail submission
Create a queue with slug 'queue2', no email fetching, same outgoing email as queue1

User sends email to email address of queue1
Fetch email via get_email management command (Ticket 'queue1-1' created)
Staff moves Ticket#1 from queue1 to queue2
Staff writes public response (Ticket in subject is 'queue2-1' from now on)
User replies to response
Fetch email via get_email management command
=> A new ticket is created because 'queue2-1' is not a valid match for queue1


Is there a reason the queue slug needs to be in email subjects?
IMO the best solution is to use a simpler format like '[Ticket#123456]'

**Comment by gwasser** [comment]
The changed queue is a good point; only the ticket number should be used internally, the slug in email subjects is for users to easily tell what type of ticket it is.

**Comment by mattbuff-nynhp** [comment]
Queue changes seem like a common use case.  So, is there a way to retain the email slug in the subject for users, yet still be robust with respect to handling queue changes? Could the regex find the slug but ignore all but the integer pk portion?

**Comment by gwasser** [comment]
Yes that's what I was referring to. I'd have to dig in to the regexes to see exactly what it is doing, but I can't think offhand of any reason why we couldn't use only the ticket number internally but still maintain the slug in email subjects.

**Comment by The-Compiler** [comment]
FWIW I don't think the queue changed in my case (there were separate mail addresses for separate queues, and I used the correct ones as far as I'm aware).

Also I think @mattbuff-nynhp also could reproduce this without changing queues, correct?

**Comment by gwasser** [comment]
Thanks! It must be some other issue, I checked the regex and the code seems to already be picking out the ticket number separately from the slug. Must be some internal logic error that will need more testing to debug.

**Comment by mattbuff-nynhp** [comment]
I can only reproduce it with changing cues.

**Comment by mattbuff-nynhp** [comment]
I just observed this bug with no queue change involved, as @The-Compiler described. The original queue slug and ticket remain in the subject, but a new ticket was opened in a different queue. The incoming email was an automatic reply. I'm not sure if that's relevant but it may be since I was unable to duplicate this bug with manually sent emails.

**Comment by gwasser** [comment]
Thanks! @mattbuff-nynhp -- if it's possible and no private info you don't mind sharing, please send the email that triggered the bug. You couldn't trigger it by resending that email? Very interesting. The original comment above has an example email too, I'm going to have to just run some examples and see if I can figure out what's going on.

**Comment by mattbuff-nynhp** [comment]
Unfortunately, I am unable send the actual email. What I've observed so far is that for some tickets, a reply results in a new ticket and for other tickets, it behaves as expected and I cannot detect a consistent difference between the two types of tickets.
I will try to debug further and will send a test case if I can re-create one.

**Comment by gwasser** [comment]
If I had to guess, it might be an encoding issue. The code base does some
funny stuff to try to detect and convert character sets, inherited from
older versions of python/django that didn't do it very well. It might be
that the encoding results in the program not detecting the correct ticket
number and therefore creating a new one.

**Comment by timthelion** [comment]
We now use a different system for identifying threads and this should be fixed.
